osu.Game/Rulesets/UI/RulesetInputManager.cs

```diff
@@ -158,69 +157,49 @@ protected override bool HandleMouseTouchStateChange(TouchStateChangeEvent e)
 
         #endregion
 
-        #region Key Counter Attachment
+        #region Component attachement
 
-        public void Attach(KeyCounterDisplay keyCounter)
-        {
-            var receptor = new ActionReceptor(keyCounter);
-
-            KeyBindingContainer.Add(receptor);
+        private readonly ActionListener actionListener;
 
-            keyCounter.SetReceptor(receptor);
-            keyCounter.AddRange(KeyBindingContainer.DefaultKeyBindings
+        public void Attach(IKeybindingListener skinComponent)
+        {
+            skinComponent.Setup(KeyBindingContainer.DefaultKeyBindings
                                                    .Select(b => b.GetAction<T>())
                                                    .Distinct()
-                                                   .OrderBy(action => action)
-                                                   .Select(action => new KeyCounterActionTrigger<T>(action)));
-        }
-
-        private partial class ActionReceptor : KeyCounterDisplay.Receptor, IKeyBindingHandler<T>
-        {
-            public ActionReceptor(KeyCounterDisplay target)
-                : base(target)
-            {
-            }
+                                                   .OrderBy(a => a));
 
-            public bool OnPressed(KeyBindingPressEvent<T> e) => Target.Counters.Where(c => c.Trigger is KeyCounterActionTrigger<T>)
-                                                                      .Select(c => (KeyCounterActionTrigger<T>)c.Trigger)
-                                                                      .Any(c => c.OnPressed(e.Action, Clock.Rate >= 0));
-
-            public void OnReleased(KeyBindingReleaseEvent<T> e)
+            if (skinComponent.CanHandleKeybindings && skinComponent is Drawable component)
             {
-                foreach (var c
-                         in Target.Counters.Where(c => c.Trigger is KeyCounterActionTrigger<T>).Select(c => (KeyCounterActionTrigger<T>)c.Trigger))
-                    c.OnReleased(e.Action, Clock.Rate >= 0);
+                try
+                {
+                    KeyBindingContainer.Add(component);
+                    return;
+                }
+                catch (Exception)
+                {
+                    return;
+                }
```

Comment:
This is again an overengineered and overfitted abstraction inferred from two separate usages, and there is no indication of any underlying structure here that would warrant defining any of this in this way. Loose sketch of how to fix this mess follows.

First of all, remove `IKeybindingEventsEmitter`. Replace it with:

```csharp
    public interface IHasKeyBindingContainer
    {
        KeyBindingContainer KeyBindingContainer { get; }
    }
```

Make `RulesetInputManager<T>` implement it. It'll just be one explicit interface implementation.

Secondly, remove _every member_ from `IKeyBindingListener` and replace with a single method:

```csharp
    public interface IKeyBindingAttachment
    {
        public void Attach(IHasKeyBindingContainer attachPoint);
    }
```

Thirdly, move _all_ setup logic of both attachable components to `IKeyBindingAttachment.Attach()`.

- In the case of the clicks per second calculator, `Attach()` should create the action listener on its own and add it to the keybinding container itself.
- In the case of the key counter, `Attach()` should read the keybindings from the container on its own and set up its children itself.

Reply:
> In the case of the key counter, Attach() should read the keybindings from the container on its own and set up its children itself.

This would mean that the `KeyCounterController` would have to get access to the generic somehow (because of `KeyCounterActionTrigger<T>`). However, both making the class and the method generic seems like a no-go since it would be called in `HUDOverlay` which has no access to that generic. 
This again means that some logic has to be implemented in `RulesetInputManager<T>` — which is afaik the only class to have access to the generic in our context — but according to your solution it should not be the case, so I'm a bit stuck right here.

Reply:
I see. Well then, it seems that the only solution available without adding more generics and more indirection - both of which are generally bad - is to revert all of the `IAttachableSkinComponent` stuff and to keep the shape of `ICanAttachHUDPieces` as it was so it can accept the `KeyCounterController` and it can accept the `ClicksPerSecondCalculator` and _nothing else_.